test/cloud-canary/mzcompose.py

```diff
@@ -66,55 +69,105 @@ def workflow_default(c: Composition) -> None:
+        td(c, "*.td")
         test_failed = False
     finally:
         # Clean up
-        print(f"Shutting down region {REGION} ...")
-        c.run("mz", "region", "disable", REGION)
+        workflow_disable_region(c)
 
     assert not test_failed
+
+
+def cloud_hostname(c: Composition) -> str:
+    print("Obtaining hostname of cloud instance ...")
+    region_status = c.run("mz", "region", "status", REGION, capture=True)
+    sql_line = region_status.stdout.split("\n")[2]
+    cloud_url = sql_line.split("\t")[1].strip()
+    cloud_hostname = urllib.parse.urlparse(cloud_url).hostname
+    return cloud_hostname
+
+
+def wait_for_cloud(c: Composition) -> None:
+    print(f"Waiting for cloud cluster to come up with username {USERNAME} ...")
+    _wait_for_pg(
+        host=cloud_hostname(c),
+        user=USERNAME,
+        password=APP_PASSWORD,
+        port=6875,
+        query="SELECT 1",
+        expected=[[1]],
+        timeout_secs=600,
+        dbname="materialize",
+        ssl_context=ssl.SSLContext(),
+        # print_result=True
+    )
+
+
+def version_check(c: Composition) -> None:
+    print("Obtaining mz_version() string from local instance ...")
+    c.up("materialized")
+    local_version = c.sql_query("SELECT mz_version();")[0][0]
+
+    print("Obtaining mz_version() string from the cloud ...")
+    cloud_cursor = pg8000.connect(
+        host=cloud_hostname(c),
+        user=USERNAME,
+        password=APP_PASSWORD,
+        port=6875,
+        ssl_context=ssl.SSLContext(),
+    ).cursor()
+    cloud_cursor.execute("SELECT mz_version()")
+    cloud_version = cloud_cursor.fetchone()[0]
+    assert (
+        local_version == cloud_version
+    ), f"local version: {local_version} is not identical to cloud version: {cloud_version}"
+
+
+def td(c: Composition, *args: str) -> None:
+    materialize_url = f"postgres://{urllib.parse.quote(USERNAME)}:{urllib.parse.quote(APP_PASSWORD)}@{cloud_hostname(c)}:6875"
+
+    with c.override(
+        Testdrive(
+            default_timeout="1200s",
+            materialize_url=materialize_url,
+            no_reset=True,  # Required so that admin port 6877 is not used
+            seed=1,  # Required for predictable Kafka topic names
+            kafka_url=KAFKA_BOOTSTRAP_SERVER,
+            schema_registry_url=SCHEMA_REGISTRY_ENDPOINT,
+            environment=[
+                "KAFKA_OPTION="
+                + ",".join(
+                    [
+                        "security.protocol=SASL_SSL",
+                        "sasl.mechanisms=PLAIN",
+                        f"sasl.username={CONFLUENT_API_KEY}",
+                        f"sasl.password={CONFLUENT_API_SECRET}",
+                    ]
+                ),
+                "VAR="
+                + ",".join(
+                    [
+                        f"confluent-api-key={CONFLUENT_API_KEY}",
+                        f"confluent-api-secret={CONFLUENT_API_SECRET}",
+                    ]
+                ),
+            ],
```

Comment:
I respectfully disagree. I am afraid we have had a few leaks already and several close calls, so using env variables would be much safer than trying to chaperone secrets using your approach through multiple pieces of code that may be written by multiple people and involve third-party services. Think of it as a continuation of the already-existing way of handling secrets from Pulumi all the way to the place where they are actually used.

Here is a list of the **known** close calls:
- testdrive was previously dumping its arguments/connection URL to stdout itself. This was removed some time ago but only because we wanted to cut on the verbosity of the output, it would have remained a leak opportunity
- buildkite is currently storing "ps ax" and "docker ps -a" outputs because we are collecting output for a sporadic Redpanda restart issue that we are experiencing in CI. If any secrets are passed as command-line arguments they can certainly end up in those existing build artifacts or any future build artifacts that contain process command lines. Your approach would **not** guard against such a leak, as the secrets would be unpacked into command-line arguments by that time. 

We have also had close calls in non-argument-related situations, so the area is known to be fraught with underwater stones:
- _wait_for_pg dumping the SQL password to stdout;
- CREATE SECRET in testdrive dumping the secret to stdout.

So I think using env variables would be the better form of defensive coding in this case.

Reply:
I don't think the environment variable approach offers much additional in the way of security, but fair enough.

> testdrive was previously dumping its arguments/connection URL to stdout itself. This was removed some time ago but only because we wanted to cut on the verbosity of the output, it would have remained a leak opportunity

That was a problem regardless of whether you the connection string was provided as an argument or an environment variable, though.

> buildkite is currently storing "ps ax" and "docker ps -a" outputs because we are collecting output for a sporadic Redpanda restart issue that we are experiencing in CI. If any secrets are passed as command-line arguments they can certainly end up in those existing build artifacts or any future build artifacts that contain process command lines. Your approach would not guard against such a leak, as the secrets would be unpacked into command-line arguments by that time.

I suppose that's true now, though I can imagine in the future we also dump `docker inspect` or something, at which point the environment variables will leak too.